Fixes #11597 - Document Request Customizers. (#11682)
eclipse/jetty.project: documentation/jetty-documentation/src/main/asciidoc/programming-guide/server/http/server-http-request-customizers.adoc

//
// ========================================================================
// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
// ========================================================================
//

[[pg-server-http-request-customizers]]
==== Request Customizers

> A request customizer is an instance of `HttpConfiguration.Customizer` and allows to customize the HTTP request and/or the HTTP response headers _before_ the `Handler` chain is invoked.

Reviewer: ```suggestion
A request customizer is an instance of `HttpConfiguration.Customizer`, that can customize the HTTP request and/or the HTTP response headers _before_ the `Handler` chain is invoked.
```

Request customizers are added to a particular `HttpConfiguration` instance, and therefore are specific to a `Connector` instance: you can have two different ``Connector``s configured with different request customizers.

For example, it is common to configure a secure `Connector` with the `SecureRequestCustomizer` that customizes the HTTP request by adding attributes that expose TLS data associated with the secure communication.

A request customizer may:

> * Process HTTP request headers, for example those injected by a proxy, and then wrap the `Request` and/or the associated `ConnectionMetaData` to transparently provide the data to applications, as if the proxy was not present.
> * Synthesize HTTP request headers from other information, for example see xref:pg-server-http-request-customizer-host[this section].
> * Wrap the `Request` object to make it secure or authenticated, if that information has been communicated via HTTP request headers, or even via out-of-band mechanisms.

Reviewer: This list is a bit too arbitrary an non-orthogonal.   How about:
```suggestion
* Inspect the received HTTP request method, URI, version and headers.
* Wrap the `Request` object to allow any method to be intercepted and customized. Typically this is done to synthesize additional HTTP request headers. 
* Set/Modify `Response` headers, including adding `PersistentHttpField`s that cannot be removed by an application reset of the response headers. 
```

Author: `PersistentHttpField` is an internal class that users have no access to.

> The out-of-the-box request customizers are:

Reviewer: ```suggestion
The out-of-the-box request customizers include:
```

* `ForwardedRequestCustomizer` -- to interpret the `Forwarded` (or the the obsolete ``+X-Forwarded-*+``) HTTP header added by a reverse proxy; see xref:pg-server-http-request-customizer-forwarded[this section].
* `HostHeaderCustomizer` -- to customize, or synthesize it when original absent, the HTTP `Host` header; see xref:pg-server-http-request-customizer-host[this section].
* `ProxyCustomizer` -- to expose as `Request` attributes the `ip:port` information carried by the PROXY protocol; see xref:pg-server-http-request-customizer-proxy[this section].
* `RewriteCustomizer` -- to rewrite the request URI; see xref:pg-server-http-request-customizer-rewrite[this section].
* `SecureRequestCustomizer` -- to expose TLS data via `Request` attributes; see xref:pg-server-http-request-customizer-secure[this section].

> You can also write your own request customizer and add it to `HttpConfiguration` instance along existing request customizers.

Reviewer: ```suggestion
You can also write your own request customizers and add them to `HttpConfiguration` instance along existing request customizers.
```

However, a very similar effect can be achieved by having the `RewriteHandler` as the first `Handler` (the child `Handler` of the `Server` instance).

Since `RewriteCustomizer` cannot send a response or complete the request/response processing, ``Rule``s that do so such as redirect rules have no effect and are ignored; only ``Rule``s that modify or wrap the `Request` will have effect and be applied.

Due to this limitation, it is often a better choice to use `RewriteHandler` instead of `RewriteCustomizer`.

For more information about how to configure `RewriteCustomizer`, see also link:{javadoc-url}/org/eclipse/jetty/rewrite/RewriteCustomizer.html[the javadocs].

[[pg-server-http-request-customizer-secure]]
===== `SecureRequestCustomizer`

`SecureRequestCustomizer` should be added when Jetty receives requests over a secure `Connector`.

`SecureRequestCustomizer` adds TLS information as request attributes, in particular an instance of `EndPoint.SslSessionData` that contains information about the negotiated TLS cipher suite and possibly client certificates, and an instance of `org.eclipse.jetty.util.ssl.X509` that contains information about the server certificate.

> `SecureRequestCustomizer` also adds, if configured so, the `Strict-Transport-Security` HTTP response header (for more information about this header, see link:https://datatracker.ietf.org/doc/html/rfc6797[its specification]).

Reviewer: We should modify that class to use a PersistentHttpField for the STS response header, so it survives any app reset

Author: I'd leave it so for now. If you really want that, we need to make `PersistentHttpField` more public, as now cannot be used outside the `jetty-server` module.

Reviewer: Perhaps we should keep PersistentHttpField internal, but after customization convert all added fields to Persistent ones.  That way the user never need know about it and no documentation required.   Different PR.
